### PR TITLE
Mask headers that match know credentials

### PR DESCRIPTION
**Pull request by timyates**
When trace logging is enabled, all headers are logged.

We should avoid this, as it may be captured by logging or build software.

We could make this configurable, if we think it's worth the extra effort.

### CODE REVIEW

**Comment by yawkat** [comment]
this could be one regex, it might be faster

**Comment by timyates** [comment]
@yawkat Good idea, done in https://github.com/micronaut-projects/micronaut-core/pull/8336/commits/bfb8960301be7c50d4073a5caf5bddd224177776

**Review by yawkat — APPROVED** [review]

File: http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java [hunk 1]
```diff
@@ -1767,17 +1772,27 @@ private void traceChunk(ByteBuf content) {
...
                 }
             } else if (!all.isEmpty()) {
-                log.trace("{}: {}", name, all.get(0));
+                String maskedValue = isMasked ? mask(all.get(0)) : all.get(0);
+                log.trace("{}: {}", name, maskedValue);
             }
         }
     }
 
+    private String mask(String value) {
```
Comment (sdelamo) [inline http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java]:
```suggestion
    @NonNull
    private String mask(@Nullable String value) {
```

**Review by sdelamo — APPROVED** [review]

File: http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java [hunk 2]
```diff
@@ -1767,17 +1772,27 @@ private void traceChunk(ByteBuf content) {
...
+        if (value == null) {
+            return null;
+        }
+        return "*MASKED* [" + value.length() + " chars]";
```
Comment (alvarosanchez) [inline http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java]: Subtle thing: I would not return the length. The fewer clues for attackers, the better :)

**Review by alvarosanchez — APPROVED** [review]

File: http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java (hunk 2 above)
Comment (timyates) [reply http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java]: I couldn't decide...  Then my pet project, I couldn't tell if the value was being injected into an auth header...

Showing the length let me see it was `Bearer ....some long token...` and not `Bearer {github.token}`

I agree though, as this feels convenient, and convenient things are usually security holes

I'll push a commit to remove it

File: http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java (hunk 1 above)
Comment (timyates) [reply http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java]: Done...  They're both nullable https://github.com/micronaut-projects/micronaut-core/pull/8336/commits/92b9c7208b11c1cef4388d66ff9fe7d1d72a5364

File: http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java (hunk 2 above)
Comment (timyates) [reply http-client/src/main/java/io/micronaut/http/client/netty/DefaultHttpClient.java]: Done https://github.com/micronaut-projects/micronaut-core/pull/8336/commits/92b9c7208b11c1cef4388d66ff9fe7d1d72a5364

**Pull request merged**